### PR TITLE
[IMM32] Rewrite ImmProcessKey

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Rewrite `ImmProcessKey` function.
- Modify `win32ss/include/imetable.h`.
- Modify `ImmProcessKey` prototype.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -4869,47 +4870,80 @@ BOOL WINAPI ImmTranslateMessage(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lKeyD
...
-            data->immKbd->uSelected--;
-            data->immKbd = new_hkl;
-            data->immKbd->pImeSelect(imc, TRUE);
-            data->immKbd->uSelected++;
+            if (LOBYTE(vKey) == VK_PACKET &&
+                !(pImeDpi->ImeInfo.fdwProperty & IME_PROP_ACCEPT_WIDE_VKEY))
+            {
+                if (pImeDpi->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+                {
+                    bCut = TRUE;
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:4897]: "cut" ? What is cut?

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -4869,47 +4870,80 @@ BOOL WINAPI ImmTranslateMessage(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lKeyD
...
+                    if (pIC->fOpen)
+                        bSkipThisKey = TRUE;
+                }
+            }
+
+            if (bUseIme)
+            {
+                if (GetKeyboardState(KeyState))
+                {
+                    vk = (bCut ? LOWORD(vKey) : vKey);
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:4911]: "cut" ? meaning of this variable is not clear at all.

File: dll/win32/imm32/imm.c [hunk 3]
```diff
@@ -4869,47 +4870,80 @@ BOOL WINAPI ImmTranslateMessage(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lKeyD
 *		ImmProcessKey(IMM32.@)
 *       ( Undocumented, called from user32.dll )
 */
-BOOL WINAPI ImmProcessKey(HWND hwnd, HKL hKL, UINT vKey, LPARAM lKeyData, DWORD unknown)
+BOOL WINAPI ImmProcessKey(HWND hWnd, HKL hKL, UINT vKey, LPARAM lParam, DWORD dwHotKeyID)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:4873]: says BOOL but the function returns flags? (the `ret` variable)

Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Fixed.

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Renamed as `bLowWordOnly`.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: Renamed as bLowWordOnly.

**Pull request merged**